File: interpreter-for-cs/src/Runtime/Sentence/SentenceExport.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

namespace peak.interpreter {

	public class SentenceExport : Sentence {
		private string _moduleName = "";
		public SentenceExport(string moduleName) {
			_moduleName = moduleName;
		}

		public override ExecuteResult Execute(Space space) {
			if (!space.SetExportModule(_moduleName)) {
				ErrorLogger.LogRuntimeError(_moduleName);
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.Export, "The module \"" + _moduleName + "\" export failed!");
				return ExecuteResult.Failed;
			}
			var module = new Module(_moduleName, space);
			if (!ModulePool.instance.AddModule(_moduleName, module, true)) {
				ErrorLogger.LogRuntimeError(_moduleName);
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.Export, "The module \"" + _moduleName + "\" export failed!");
				return ExecuteResult.Failed;
			}
			return ExecuteResult.Successed;
		}

	}

} // namespace peak.interpreter

File: interpreter-for-cpp/src/Runtime/Sentence/SentenceExport.cpp
#include "SentenceExport.h"
#include "../Module.h"
#include "../ModulePool.h"

using namespace peak::interpreter;

SentenceExport::SentenceExport(const std::string& moduleName)
	: _moduleName(moduleName) {
}

ExecuteResult SentenceExport::Execute(std::shared_ptr<Space> space) {
	if (space->GetSpaceType() != SpaceType::None) {
		ErrorLogger::LogRuntimeError(_moduleName);
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::Export, "The module \"" + _moduleName + "\" can't export here!");
		return ExecuteResult::Failed;
	}
	if (!space->SetExportModule(_moduleName)) {
		ErrorLogger::LogRuntimeError(_moduleName);
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::Export, "The module \"" + _moduleName + "\" already export!");
		return ExecuteResult::Failed;
	}
	auto module = std::shared_ptr<Module>(new Module(_moduleName, space));
	if (!ModulePool::GetInstance()->AddModule(_moduleName, module, true)) {
		ErrorLogger::LogRuntimeError(_moduleName);
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::Export, "The module \"" + _moduleName + "\" already exists!");
		return ExecuteResult::Failed;
	}
	return ExecuteResult::Successed;
}